Gradle improvements v2

mahtaran opened the pull request:
Organised the `.gitignore` file and added directories which should be ignored.
Updated to Gradle 4.4 from 2.0. Changed several things in accordance with this.
Moved all configurable settings from the `build.gradle` file to `build.properties`.
Increased the maximum amount of RAM Gradle can use while compiling for faster builds.
Added all required libs (and some not required at this moment but they might be in the future (SuffixForge)) to build the mod.
Added an automated system which replaces placeholders like the modid, version and Minecraft version.

After it is reviewed, it should be tested if the newly compiled JAR doesn't break everything.

mahtaran committed:
Gradle improvements
Improved the .gitignore
Updated to Gradle 4.4 from 2.0. Changed several things in accordance with this
Moved all configurable settings from the build.gradle file to build.properties
Increased the maximum amount of RAM Gradle can use while compiling for faster builds
Added all required libs (and some not required at this moment but they might be in the future (SuffixForge)) to build the mod
Added an automated system which replaces placeholders like the modid, version and Minecraft version

Signed-off-by: Mahtaran <mahtaran@amuzil.com>

CraftedMods reviewed (changes requested):
That is the first part of my review.

mahtaran committed:
Ignored the `libs` folder

Mahtaran commented on build.gradle:
```diff
@@ -1,83 +1,260 @@
+    provided "org.bukkit:craftbukkit-remapped:1.7.10"
+
+
+    shadow "io.github.lukehutch:fast-classpath-scanner:2.21"
```
What do you mean? I’m pretty sure it was in the compiled JAR.

CraftedMods replied:
I did build it several times, and it wasn't there.

Mahtaran replied:
using what task did you build?

CraftedMods replied:
`build`

Mahtaran replied:
Problem confirmed. Moving to fix it now.

Mahtaran replied:
@CraftedMods Should be resolved. Feel free to try again.

CraftedMods replied:
I'll try it tomorrow.

mahtaran committed:
Fixed the `build` task not including shadowed dependencies

CraftedMods commented on build.gradle:
```diff
@@ -1,83 +1,261 @@
+
+    // Forge Mods
+    if (dependenciesConfig.obfuscated) {
+        provided "com.enderio.core:endercore:0.2.0-beta.39"
```
Why don't we just use the `compile` configuration?

Mahtaran replied:
Because I thought using `provided` as keyword would make more sense. I have replaced it with `compile` in my latest commit.

CraftedMods commented on build.gradle:
```diff
@@ -1,83 +1,261 @@
 
 configurations {
-    exports
+    shadow
```
Innecessary line. The shadow plugin registers the configuration by it's own.

CraftedMods commented on build.gradle:
```diff
@@ -1,83 +1,261 @@
+    provided "net.citizensnpcs:citizens:2.0.13-alpha.1"
+    provided "org.bukkit:craftbukkit-remapped:1.7.10"
+
+
```
Please replace that empty line with a comment.
```suggestion
// Dependencies that will be included ("shadowed") into the built JAR
```

CraftedMods commented on build.gradle:
```diff
@@ -1,83 +1,261 @@
+    replace "@name@", config.name
 }
 
 jar {
```
The `jar` task can be completely removed, because the `shadowJar` task extends it.
You can replace the current `jar` and `shadowJar` tasks with that code: ```
shadowJar {
	classifier= ""
	
    manifest {
        attributes "FMLCorePlugin": "pvpmode.internal.common.core.PvPModeCore"
        attributes "FMLCorePluginContainsFMLMod": "true"
    }
    
    from("./") {
        include "*.md" // Include the markdown files
    }
    
    exclude "META-INF/**"
    exclude "META-INF/*"
    
    configurations = [project.configurations.shadow]
}
```

Mahtaran replied:
Why do you exclude the META-INF files? The `shadow` task automatically removes these from shadowed dependencies.

CraftedMods replied:
I thought that too, but then I rechecked the documentation - it only removes those:
`META-INF/INDEX.LIST`
`META-INF/*.SF`
`META-INF/*.DSA`
`META-INF/*.RSA`

CraftedMods commented on build.gradle:
```diff
@@ -1,83 +1,261 @@
 dependencies {
-    exports 'io.github.lukehutch:fast-classpath-scanner:2.21'
-    configurations.compile.extendsFrom(configurations.exports)
+    provided fileTree(dir: dependenciesConfig.libraries + (dependenciesConfig.obfuscated ? "" : "Deobf"), include: ["*.jar"])
```
What does that do?

Mahtaran replied:
This adds all files in the `libs` or `libsDeobf` to the compile class path.

CraftedMods committed:
Update build.gradle

Co-Authored-By: Mahtaran <mahtaran@amuzil.com>

Mahtaran commented on build.gradle:
```diff
@@ -1,83 +1,261 @@
+    }
+}
+
+task cleanBuild(dependsOn: ["buildInformation", "startBuild", "clean", "build", "shadowJar", "finishBuild"]) {
```
Gradle recommends that you don’t change the logic of built-in tasks as it may result in hard to find bugs.
I added the `clean` task

CraftedMods replied:
I'm not completely sure, but isn't the `cleanBuild` task slower than the `build` task, because really everything will be rebuilt? If that's the case, I would suggest you to move the `clean` part to the `publish` task. If one wants a `clean` environment, one can always use `gradle build clean`.

Mahtaran replied:
I’ll look into it.

Mahtaran replied:
After doing 5 consecutive runs of both tasks, I got the following results:

> ```gradlew cleanBuild```
> 9s, 7s, 7s, 7s, 7s
> 16 actionable tasks: 14 executed, 2 up-to-date
> 
> ```gradlew build```
> 2s, 2s, 2s, 1s, 2s
> 12 actionable tasks: 4 executed, 8 up-to-date

It seems like you're right. I'll merge this task with the `publish` task.

Mahtaran commented:
The rest will be addressed tomorrow when I have access to a PC again.

mahtaran committed:
Cleaned up `build.gradle`
Removed the `configurations` block. Now uses the `compile` configuration instead of `provided`.
Merged the `jar` and `shadowJar` task.
Relocated `fastclasspathscanner` from `io.github.lukehutch.fastclasspathscanner` to `pvpmode.fastclasspathscanner`.
Removed the `reobf` block.
Tested with `cleanBuild` and `build`, `fastclasspathscanner` is included and relocated in both, imports are adjusted for the relocation in both, obfuscation of source still happens without `reobf` block in both.

Signed-off-by: Mahtaran <mahtaran@amuzil.com>

CraftedMods commented on build.gradle:
```diff
@@ -1,83 +1,248 @@
+	classifier = ""
+	configurations = [project.configurations.shadow]
+
+	relocate "io.github.lukehutch.fastclasspathscanner", "pvpmode.fastclasspathscanner"
```
For "credit" and code separation reasons, I think it would be better to name it `pvpmode.shadow.io.github.lukehutch.fastclasspathscanner`.

CraftedMods commented on build.gradle:
```diff
@@ -1,83 +1,248 @@
+	}
+}
+
+def startBuildMessage = "Starting build...\n\nName: ${project.name}\nVersion: ${project.version}"
```
We could replace the whole build message code with that:
```
gradle.afterProject {
    println "\n### Building ${project.name} (${project.version}) ###\n"
}
```
The message could even be colored, if we would use internal classes.

Mahtaran replied:
I primarily added this since we use a similar message on another mod I’m working on, but there it’s mostly because the mod is aimed to be built by users without a lot of technical knowledge. I guess that we could remove the last message and make the first shorter.

CraftedMods committed:
Update build.gradle

Co-Authored-By: Mahtaran <mahtaran@amuzil.com>

Mahtaran commented on build.gradle:
```diff
@@ -1,83 +1,248 @@
+}
+
+def startBuildMessage = "Starting build...\n\nName: ${project.name}\nVersion: ${project.version}"
+def finishBuildMessage = "Successfully finished building ${project.name} version ${project.version}\n\nYour build is located in:\nbuild/libs"
```
Again, this is primarily something we did on my other project to help users who do not know where Gradle stores the built files.

CraftedMods committed:
Update build.gradle

Co-Authored-By: Mahtaran <mahtaran@amuzil.com>

Mahtaran commented on build.gradle:
```diff
@@ -1,83 +1,248 @@
+	compile fileTree(dir: dependenciesConfig.libraries + (dependenciesConfig.obfuscated ? "" : "Deobf"), include: ["*.jar"])
+
+	// Forge Mods
+	if (dependenciesConfig.obfuscated) {
```
I guess that’d be possible.

mahtaran committed:
Addressed several issues
Added dynamic appendage of the "-deobf" suffix to dependencies marked as being obfuscated if in a deobfuscated environment
Removed the message after the build finished. Simplified the message at the start.
Removed the `cleanBuild` task.

Signed-off-by: Mahtaran <mahtaran@amuzil.com>

mahtaran committed:
Merge remote-tracking branch 'origin/mahtaran/gradle-improvements' into mahtaran/gradle-improvements

# Conflicts:
#	build.gradle

mahtaran committed:
Small change to my repository

mahtaran committed:
Fixed 2 small errors
The method didn't work due to it being static and accessing a local variable. Fixed by not making it static.
The repository didn't work if you didn't provide credentials. Added the public repo separately to circumvent this.

Signed-off-by: Mahtaran <mahtaran@amuzil.com>

CraftedMods commented on build.gradle:
```diff
@@ -1,83 +1,224 @@
+	}
+}
+
+tasks.wrapper.dependsOn buildInformation
```
When is that executed?

Mahtaran commented:
Wrapper is the first task executed in the build, haven’t tested if it’s
executed in other tasks, but that doesn’t really matter in my opinion, as
all executed tasks have to do with the project.

On Thu, 14 Feb 2019 at 10:40, CraftedMods <notifications@github.com> wrote:

> *@CraftedMods* commented on this pull request.
> ------------------------------
>
> In build.gradle
> <https://github.com/Lembas-Modding-Team/pvp-mode/pull/330#discussion_r256755221>
> :
>
> > +		for (String line : lines)
> +			maxLength = Math.max(line.length(), maxLength)
> +		for (def i = 0; i < maxLength; i++) {
> +			border += i % 2 == 0 ? "=" : "-"
> +		}
> +		println """
> +${border}
> +
> +${message}
> +
> +${border}
> +"""
> +	}
> +}
> +
> +tasks.wrapper.dependsOn buildInformation
>
> When is that executed?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Lembas-Modding-Team/pvp-mode/pull/330#pullrequestreview-203646024>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVrKm4QpezzSOb-qWU7tnintIY5EinjCks5vNS8XgaJpZM4aNG4a>
> .
>

CraftedMods commented:
The task isn't executed. I tried using both `build` and `clean build`.

Mahtaran commented:
> 
> 
> The task isn't executed. I tried using both `build` and `clean build`.

You are using the `gradlew` command instead of the `gradle` command right?

CraftedMods commented:
Yes.

CraftedMods commented:
Also, I noticed, that I cannot use the Hot Code replacement eclipse has, because I've to start the program via gradle now. That's bad and makes development hard, for some things, because then I've to restart the program with every change. It's probably still possible somehow, but not as easy as it was.

CraftedMods commented:
If you don't know a better solution, I would recommend to remove these lines:
```
replaceIn "src/main/java/pvpmode/PvPMode.java"
	replace "@version@", config.version
	replace "@modid@", config.modid
	replace "@mcversion@", project.minecraft.version
	replace "@name@", config.name
```
This is not ideal, but honestly I rather want to change  the mod version at one place more before an update, than not having the hotswapping. Starting the program without gradle is also more convenient and more easy, for me in eclipse, because the Forge workspace layout creates some problems with the eclipse gradle plugin.

Mahtaran commented:
I guess we could work around 

> 
> 
> Yes. `gradle` doesn't work, using the Forge Dev Environment.

Ah, problem confirmed. Weird, it worked before. Resolved, replaced `tasks.wrapper` with `tasks.compileApiJava`. Going to be in the next commit, after I look into the hot swapping.
